### ISSUE TITLE
CastError Model vs Data validation (with test)

### ISSUE THREAD

**Issue by llafuente** [LearnBoost/mongoose#2767]
When you have a value that it's not the same type as the one defined in the model, mongoose raise a CastError, even if you fix the type.

I think the reason is because mongoose try to use the validation from the data got from mongo and not the one in the model.

I search in the forum there are maybe two related issues:
https://github.com/LearnBoost/mongoose/issues/2766 (String to objectId)
https://github.com/LearnBoost/mongoose/issues/1865

This is the test: Number -> String
```js
var mongoose = require('mongoose'),
    Schema = mongoose.Schema;

var a_schema = new Schema({
    value: Number
},{
  collection: "test"
});

var b_schema = new Schema({
    value: String
},{
  collection: "test"
});

var a_model = mongoose.model('a', a_schema);
var b_model = mongoose.model('b', b_schema);

mongoose.connect("mongodb://localhost/test");

mongoose.connection.on('open', function (ref) {
  // create a -> 100(number)
  a_model.create({
    value: 100
  }, function(err, data) {
    console.log("create a err", err);
    console.log("create a data", data);

    // get created a as b (value is 100 number)
    b_model.findById(data._id, function() {
      console.log("get as b data", data);
      console.log("get as b err", err);

      // change b.value to a string
      data.value = "string";
      data.save(function(err, data) {
        console.log("save as b err", err); // <-- CastError bug
        console.log(err.stack);
        console.log("save as b data", data);

        // b can be created with a string...
        b_model.create({
          value: "string"
        }, function(err, data) {
          console.log("create b err", err);
          console.log("create b data", data);
        });
      });
    });
  });
});
```

**Comment by vkarpov15** [comment]
Not a bug, in this case `data` is an instance of `a_model` in the line that has `// <-- CastError bug`. #2766 is something that I thought was fixed in 4.0 but turns out isn't. Will put up a fix for that shortly.

**Comment by llafuente** [comment]
sorry forgot the arguments.

I tried to isolate the code from out production env, we have the CastError with something very similiar, will see if i found a proper test.

**Comment by vkarpov15** [comment]
If it has to do with stale cast errors, 4.0.0 won't have that issue any longer :)

**Comment by llafuente** [comment]
confirmed it's solved @4.0.7
The field will be null in this case after find*, because cast failed... ok!
But if you use `lean` it will be the original.. it's a bit noisy but it's how it works.
